x-pack/plugins/security/server/session_management/session.ts

```diff
@@ -358,24 +370,54 @@ export class Session {
   }
 
   /**
-   * Clears session value for the specified request.
-   * @param request Request instance to clear session value for.
+   * Invalidates sessions that match the specified filter.
+   * @param request Request instance initiated invalidation.
+   * @param filter Filter that narrows down the list of the sessions that should be invalidated.
    */
-  async clear(request: KibanaRequest) {
+  async invalidate(request: KibanaRequest, filter: InvalidateSessionsFilter) {
+    // We don't require request to have the associated session, but nevertheless we still want to
+    // log the SID if session is available.
     const sessionCookieValue = await this.options.sessionCookie.get(request);
-    if (!sessionCookieValue) {
-      return;
-    }
+    const sessionLogger = this.getLoggerForSID(sessionCookieValue?.sid);
 
-    const sessionLogger = this.getLoggerForSID(sessionCookieValue.sid);
-    sessionLogger.debug('Invalidating session.');
+    let invalidateIndexValueFilter;
+    if (filter.match === 'current') {
+      if (!sessionCookieValue) {
+        return;
+      }
 
-    await Promise.all([
-      this.options.sessionCookie.clear(request),
-      this.options.sessionIndex.clear(sessionCookieValue.sid),
-    ]);
+      sessionLogger.debug('Invalidating current session.');
+      await this.options.sessionCookie.clear(request);
+      invalidateIndexValueFilter = { match: 'sid' as 'sid', sid: sessionCookieValue.sid };
+    } else if (filter.match === 'all') {
+      sessionLogger.debug('Invalidating all sessions.');
+      await this.options.sessionCookie.clear(request);
+      invalidateIndexValueFilter = filter;
+    } else {
+      sessionLogger.debug(
+        `Invalidating sessions that match query: ${JSON.stringify(
+          filter.query.username ? { ...filter.query, username: '[REDACTED]' } : filter.query
+        )}.`
+      );
+
+      // We don't clear session cookie in this case since we cannot always be sure that the query
+      // will match the current session. This behavior doesn't introduce any risk since even if the
+      // current session has been affected the session cookie will be automatically invalidated as
+      // soon as client attempts to re-use it due to missing underlying session index value.
+      invalidateIndexValueFilter = filter.query.username
+        ? {
+            ...filter,
+            query: {
+              provider: filter.query.provider,
+              usernameHash: createHash('sha3-256').update(filter.query.username).digest('hex'),
```

Comment:
nit: this is the third place where we're calculating the username hash. This might be a good opportunity to consolidate the logic into a single function.

Reply:
Good catch! Will move to a separate function.